src/operator/tensor/matrix_op-inl.h

```diff
@@ -476,6 +476,141 @@ void DotBackward_(const nnvm::NodeAttrs& attrs,
   }
 }
 
+template<bool transpose>
+struct DotCsrDense;
+
+/*!
+ * \brief dot(csr, dense)
+ */
+template<>
+struct DotCsrDense<false> {
+  /*!
+   * \brief This function represents performing an inner product between a row of lhs
+   * and a column of rhs and then assigning the value to out[i].
+   * \param i i-th element in out 1D view
+   * \param out output matrix
+   * \param data_l csr values of lhs
+   * \param indptr_l csr indptr of lhs
+   * \param col_idx_l csr col_idx of lhs
+   * \param data_r dense data of rhs
+   * \param num_rows number of rows of output
+   * \param num_cols number of columns of output
+   */
+  template<typename DType, typename IType, typename CType>
+  MSHADOW_XINLINE static void Map(int i, DType* out, const DType* data_l, const IType* indptr_l,
+                                  const CType* col_idx_l, const DType* data_r,
+                                  const int num_rows, const int num_cols) {
+    const int irow = i / num_rows;  // row id of the lhs
+    const int icol = i % num_cols;  // col id of the rhs
+    out[i] = 0;
+    for (IType j = indptr_l[irow]; j < indptr_l[irow+1]; ++j) {
+      const CType cur_col = col_idx_l[j];  // corresponding row id of the rhs
+      out[i] += data_l[j] * data_r[cur_col*num_cols+icol];
+    }
+  }
+};
+
+/*!
+ * \brief dot(csr.T(), dense)
+ */
+template<>
+struct DotCsrDense<true> {
+  /*!
+   * \brief This function represents performing an inner product between a column of lhs
+   * and a column of rhs and then assigning the value to out[i].
+   * \param i i-th element in out 1D view
+   * \param out output matrix
+   * \param data_l csr values of lhs
+   * \param indptr_l csr indptr of lhs
+   * \param col_idx_l csr col_idx of lhs
+   * \param data_r dense data of rhs
+   * \param num_rows_l number of rows of lhs
+   * \param num_rows number of rows of output
+   * \param num_cols number of columns of outputs
+   */
+  template<typename DType, typename IType, typename CType>
+  MSHADOW_XINLINE static void Map(int i, DType* out, const DType* data_l, const IType* indptr_l,
+                                  const CType* col_idx_l, const DType* data_r, const int num_rows_l,
+                                  const int num_rows, const int num_cols) {
+    const int irow = i / num_rows;  // col id of the lhs
+    const int icol = i % num_cols;  // col id of the rhs
+    out[i] = 0;
+    for (int i = 0; i < num_rows_l; ++i) {
+      const IType low = indptr_l[i];
+      const IType high = indptr_l[i+1];
+      if (low == high || irow < col_idx_l[low] || irow > col_idx_l[high-1]) continue;
+      int j = low;
+      while (j < high) {
+        if (irow == col_idx_l[j]) {
+          out[i] += data_l[j] * data_r[i*num_cols+icol];
+          break;
+        }
+        ++j;
+      }
+    }
+  }
+};
+
+template<typename xpu>
+inline void DotForwardCsrDense(const nnvm::NodeAttrs& attrs,
```

Comment:
Let's always use OpEx when registering with nnvm. DotForwardEx would internally either fallback, or call OpCsrDn, OpCsrCsr, etc. Removing the Ex from DotForwardExCsrCsr since it already implies sparse ndarray